polaris-react/src/components/IndexTable/IndexTable.scss

```
@@ -232,6 +276,11 @@ $loading-panel-height: 53px;
   font-size: var(--p-font-size-75);
   white-space: nowrap;
   border: 0;
+
+  #{$se23} & {
+    padding: var(--p-space-1_5-experimental);
+    line-height: var(--p-font-line-height-1);
+  }
```

You can modify the right-side padding for TableHeading here so that it matches the 16px left padding for the first TableHeading. IndexTable isn't set up to pinpoint the `-last` heading like we do the `-first` and `-second` so I think it should be okay to use `:last-child` otherwise it'll be a lot more changes on the tsx file. 

```suggestion

    &:last-child {
      padding-right: var(--p-space-3);
    }
  }
```